---
title: "*TBXT* chromatin looping analysis in lung cancer"
author: "<h4>Authors: <i>Reuben M. Yaa</i></h4>"
date: "<h4>README updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output: github_document
---
## Introduction
While high levels of TBXT have been detected in metastatic lung cancer, we used circularised chromosome capture combined with sequencing (4C-seq) to characterise if physical genome wide *cis* -interactions  differences around *TBXT* loci in (1) lung cancer cells (H460), a TBXT expressing cell line and
(2) lung cancer cells (H358 and A549), non-TBXT expressers have association with TBXT regulation


## 4C template and library
A543, H358 and H460 4C templates and libraries were prepared in replicates using *Nla*III and *Dpn*II restriction enzymes and sequenced at 150bp PE

## 4C data analysis
Mapping

Reads were processed using [4C pipeline](https://github.com/mangiruben/TBXT_regulation/blob/main/code/4C_pipeline.pl)

Interaction landscape calling

Statistically interacting fragments were called using peakC embedded in [interact_calling pipeline](https://github.com/mangiruben/TBXT_regulation/blob/main/code/interact_calling.R)

## Downstream analysis
Analysis of the statistically interacting fragments

The fragments now called loops were analysed in different ways as follows;

  - Venn diagrams for comparing loops between *TBXT* expression conditions
  - Merged loop for visualisation of the regulation landscape linear genome using [spider_plots](https://github.com/mangiruben/TBXT_regulation/blob/main/code/spider_plot.R)
  - Clustering of loops as shared, gained or lost using Bedtools suit
  - Density analysis of clustered loops done using [4C_changing density](https://github.com/mangiruben/TBXT_regulation/blob/main/code/4C_density.Rmd)
  - Motif analysis of gained loops using [RSAT](http://rsat.sb-roscoff.fr/) 
    
Epigenome enrichment analysis

ChIP-seq analysis of H358, A549 and H460 publicly available datasets using [ChIP_pipeline](https://github.com/mangiruben/TBXT_regulation/blob/main/code/chip_seq_SE.pl). 
Track collections are saved in UCSC [lung_TBXT](https://genome-euro.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr6%3A166149847%2D166189172&hgsid=290115163_VLmnkvTVt1edA6syLcggfafmtdXD)

<hr>